artiq/coredevice/urukul.py

```diff
@@ -124,6 +135,490 @@ def set_mu(self, ftw: TInt32):
+
+
+class ProtoRev9(CPLDVersion):
+    """
+    Implementation of the CPLD for Urkul ProtoRev9.
+    """
+
+    # ProtoRev9 CFG configuration register bit offsets
+    CFG_OSK = 20
+    CFG_DRCTL = 24
+    CFG_DRHOLD = 28
+    CFG_IO_UPDATE = 32
+    CFG_MASK_NU = 36
+    CFG_CLK_SEL0 = 40
+    CFG_CLK_SEL1 = 44
+    CFG_SYNC_SEL = 41
+    CFG_RST = 42
+    CFG_IO_RST = 43
+    CFG_CLK_DIV = 45
+    CFG_ATT_EN = 47
+
+    @staticmethod
+    @portable
+    def urukul_cfg(
+        rf_sw,
+        led,
+        profile,
+        osk,
+        drctl,
+        drhold,
+        io_update,
+        mask_nu,
+        clk_sel,
+        sync_sel,
+        rst,
+        io_rst,
+        clk_div,
+        att_en,
+    ):
+        """Build Urukul CPLD configuration register"""
+        return (
+            (rf_sw << CFG_RF_SW)
+            | (led << CFG_LED)
+            | (profile << CFG_PROFILE)
+            | (osk << ProtoRev9.CFG_OSK)
+            | (drctl << ProtoRev9.CFG_DRCTL)
+            | (drhold << ProtoRev9.CFG_DRHOLD)
+            | (io_update << ProtoRev9.CFG_IO_UPDATE)
+            | (mask_nu << ProtoRev9.CFG_MASK_NU)
+            | ((clk_sel & 0x01) << ProtoRev9.CFG_CLK_SEL0)
+            | ((clk_sel & 0x02) << (ProtoRev9.CFG_CLK_SEL1 - 1))
+            | (sync_sel << ProtoRev9.CFG_SYNC_SEL)
+            | (rst << ProtoRev9.CFG_RST)
+            | (io_rst << ProtoRev9.CFG_IO_RST)
+            | (clk_div << ProtoRev9.CFG_CLK_DIV)
+            | (att_en << ProtoRev9.CFG_ATT_EN)
+        )
+
+    @kernel
+    def cfg_write(self, cpld, cfg: TInt64):
+        """Write to the configuration register.
+
+        See :func:`urukul_cfg` for possible flags.
+
+        :param cfg: 52-bit data to be written. Will be stored at
+            :attr:`cfg_reg`.
+        """
+        cpld.bus.set_config_mu(SPI_CONFIG, 24, SPIT_CFG_WR, CS_CFG)
+        cpld.bus.write(((cfg >> 28) & 0xFFFFFF) << 8)
+        cpld.bus.set_config_mu(SPI_CONFIG | spi.SPI_END, 28, SPIT_CFG_WR, CS_CFG)
+        cpld.bus.write((cfg & 0xFFFFFFF) << 4)
+        cpld.cfg_reg = cfg
+
+    @kernel
+    def sta_read(self, cpld) -> TInt64:
+        """Read the status register.
+
+        Use any of the following functions to extract values:
+
+            * :func:`urukul_sta_rf_sw`
+            * :func:`urukul_sta_smp_err`
+            * :func:`urukul_sta_pll_lock`
+            * :func:`urukul_sta_ifc_mode`
+            * :func:`urukul_sta_proto_rev`
+            * :func:`urukul_sta_drover`
+
+        :return: The status register value.
+        """
+        cpld.bus.set_config_mu(SPI_CONFIG | spi.SPI_INPUT, 24, SPIT_CFG_RD, CS_CFG)
+        cpld.bus.write(((cpld.cfg_reg >> 24) & 0xFFFFFF) << 8)
+        cpld.bus.set_config_mu(
+            SPI_CONFIG | spi.SPI_END | spi.SPI_INPUT, 28, SPIT_CFG_RD, CS_CFG
+        )
+        cpld.bus.write((cpld.cfg_reg & 0xFFFFFFF) << 4)
+        hi = cpld.bus.read()
+        lo = cpld.bus.read()
+        return (int64(hi) << 28) | lo
```

Comment:
The SR has 27 bits only. What is the significance of reporting `hi` as well instead of just discarding after read?

Reply:
Actually, I think you can set the config of the first SPI transaction to not generate an RTIO input in the first place.

Reply:
> Actually, I think you can set the config of the first SPI transaction to not generate an RTIO input in the first place.

I am not sure what you mean here.  Can you elaborate?

Reply:
Is it possible to not set SPI_INPUT flag for the first 24-bits SPI transaction, and leave the second 28-bits transaction as-is (with SPI_INPUT)? If you do this, the RTIO core should only generate an input for the second 28-bits transaction only.